content-testing/validators.js

```diff
@@ -0,0 +1,285 @@
+
+      if (value !== `${year}-${month}-${day}`) correctFormat = false;
+    } else {
+      // ISO 8601 timestamp
+      if (value !== date.toISOString()) correctFormat = false;
+    }
+
+    if (!correctFormat) {
+      return context.createError({
+        message: `Date is not correctly formatted`
+      });
+    }
+
+    // verify that the date is within defined bounds
+    if (date < earliest) {
+      return context.createError({
+        message: `Date is before ${earliest.toLocaleString()}")`
+      });
+    }
+    if (date > now) {
+      return context.createError({ message: `Date is in the future` });
+    }
+
+    return true;
+  }
+);
+
+const urlRegex = /^http[s]?:\/\//;
+const nonUrlStringValidator = string().test(
+  'non-URL string',
+  'String cannot be a URL',
+  (value) => {
+    if (!value) return true;
+    return !urlRegex.test(value);
+  }
+);
+
+const youtubeIdRegex = /^[A-Za-z0-9_\-]{11}$/;
+const youtubeIdValidator = string().test(
+  'YouTube ID',
+  'Should be a valid YouTube ID',
+  (value) => {
+    if (!value) return true;
+    return youtubeIdRegex.test(value);
+  }
+);
+
+// this 12 chars prefix seems to reliably identify the Coding Train channel
+const youtubePlaylistIdRegex = /^PLRqwX-V7Uu6[A-Za-z0-9_\-]{22}$/;
+const youtubePlaylistIdValidator = string().test(
+  'YouTube playlist ID',
+  'Should be a valid YouTube playlist ID from the Coding Train channel',
+  (value) => {
+    if (!value) return true;
+    return youtubePlaylistIdRegex.test(value);
+  }
+);
+
+const videoNumberRegex = /^C?\d+(?:\.\d+)?$/;
+const videoNumberValidator = string().test(
+  'video number',
+  'A video number should be a number optionally prefixed with `C`',
+  (value) => {
+    if (!value) return true;
+    return videoNumberRegex.test(value);
+  }
+);
+
+// 00:00 to 59:59, 0:00:00 to 9:59:59 - optional leading zeros for minutes and seconds
+const timestampRegex = /^(?:[0-9]:)?[0-5]?\d:[0-5]?\d$/;
+const timestampValidator = string().test(
+  'timestamps',
+  'A timestamp should be in the `mm:ss` or `h:mm:ss` format',
+  (value) => {
+    if (!value) return true;
+    return timestampRegex.test(value);
+  }
+);
+
+const timestampToSeconds = (timestamp) => {
+  // NOTE: copied from `./gatsby/utils.mjs`, can't use directly for now because of ESM/CJS mismatch
+
+  return timestamp
+    .split(':')
+    .reverse()
+    .map(Number)
+    .reduce((totalSeconds, component, index) => {
+      return totalSeconds + component * 60 ** index;
+    }, 0);
+};
+
+const timestampsArrayValidator = array(
+  strictObject({
+    time: timestampValidator.required(),
+    title: string().required()
+  }).required()
+).test((values, context) => {
+  if (!values) return true;
+
+  // sequential timestamps
+  let errors = [];
+  let previousTime = -1;
+
+  for (const value of values) {
+    const currentTime = timestampToSeconds(value.time);
+    if (currentTime <= previousTime) errors.push(value);
+    previousTime = currentTime;
+  }
```

Comment:
Another possible test may be to check if the first timestamp is `0:00` (or `00:00`/`0:00:00`), or else youtube doesn't parse the timestamps as chapters.

Reply:
Yes, I had coded this at some point but removed it for some reason. Maybe thinking it was too strict? I did not realize it would break YouTube chapters, sounds like it should be added back in.

Reply:
I corrected 5 videos that were missing `0:00` timestamp, you can see them in [94e0c03](https://github.com/CodingTrain/thecodingtrain.com/pull/1246/commits/94e0c03750ab90a1f3b292f89b8e776d05067076)

Reply:
I also found documentation on [YouTube video chapters](https://support.google.com/youtube/answer/9884579?hl=en), we can also validate that there are at least 3 timestamps per video (if any), and that they are at least 10 seconds between each other.